Config File Switching

Ligraph opened the pull request:
WIP.

Features I want:
- [x] Can switch config files and thumbnails load properly
- [x] Program remembers what your last config file was and loads it
- [x] Create a default config file if none exist in the directory
- [x] Copy config files from UI
- [x] Delete config files from UI (need are you sure dialog)
- [x] Rename config file names from UI (use a dialog for name?)

Also need a better looking UI.

Ligraph committed:
Basic config file switching implemented.

Ligraph committed:
Made the default config file load properly when no config files are present.

Ligraph committed:
Fixed creating new files when the last used config file has been deleted.

Phrynohyas commented:
I have to ask you to redo the pull request :(

Iuuse is that you've commited a lot of packages that shouldn't be in the repository. Cleaning them out if I'll merge this branch will be a mess

Which Git client do you use that allows to commit way to much files?

Ligraph commented:
I've been using the addon for Visual Studio.  I think I forgot to add packages to the gitignore.

Ligraph committed:
Updated my gitignore

Ligraph committed:
Removed packaged.

Ligraph commented:
The extra files are fixed (they shouldn't be part of the pull request any more), do you still need me to create a new pull request?

Phrynohyas commented:
Not needed.

Actually I'd like to discuss with you the following idea - to make the config selection code and GUI a part of a separate dialog (for now it can be opened via button press in the main dialog)

Thus the code will be separated and the main GUI will be keep clean (actually I think the main gialog is overcomplicated already for an average user, so adding even more controls won't help)

Ligraph commented:
I like that idea, I'll see what I can figure out.  For now its finals week so I probably won't get much done in the next few days.

What if instead of a whole separate dialog the main window just expanded to the side?  I'm not sure how that would work in the code, but it seems like it would be a little cleaner UI wise.  And we could put other controls there eventually.

Phrynohyas commented:
Separate dialog is better because this functionality (config switching) is completely different from the rest of the main dialog options.
I'll eventually move that dialog to the tray menu so it will be completely separated from the main options dialog

Ligraph committed:
Added a config settings window, and implenented config renaming

Ligraph committed:
Added file renaming, copying, and deleting.
Current bugs:  deleting current file does weird things with names, renaming current config file doesn't update properly.

Ligraph committed:
Renaming the current config file's filename now updates everything properly.

Ligraph committed:
Fixed deleting files doing odd things with names.

Ligraph commented:
I created a separate dialog for renaming, moving, copying, and deleting.  I left the dropdown in because it is small, and I think having to open and click through a separate dialog just to change configs would be a pain.
UI so far:
![image](https://cloud.githubusercontent.com/assets/17717986/21473678/a4ba804a-cac8-11e6-9eee-c3ad65f8d4c9.png)


I still need to add the rightclick tray menu and streamling the config file renaming (remove config/ and .json).

Ligraph committed:
Fixed file UI to not show config/ and .json.

Ligraph committed:
Added the tray menu and fixed a few bugs.

Ligraph committed:
Cleaned up some debuging output.

New bug:  If you delete default, the next config deleted will be replaced by default.

Ligraph committed:
Fixed a few more default-related bugs.

Ligraph committed:
Renamed a devider.

Ligraph commented:
Need to test that config files load correctly with deleted/renamed events.

Ligraph committed:
Fixed a few bugs.  Previews still don't reset to original positions properly.

Ligraph commented:
So, if you delete your last config, thumbnail positions will **not** reset.  This is not intended, but occurs because in RefreshThumbnails and UpdateThumbnailsList when GetThumbnailLocation is called the current thumbnail position is passed as the default, thus, since the newly generated config is blank, the current positions are used.  Note that they are not saved because they are never written to the config file, unless someone moves or resizes them.

As far as I can tell everything else is working properly, I'm going to try to handle this in the next few days, but I'm not yet sure of the best way.  Unless you have any objections I plan on just using the default position on startup (whatever/wherever that is).

Edit:  This also means that if a thumbnail position is not included in a config it will automatically go to the default position instead of its current position when the config is loaded, which I think is a positive (It mimics the current behavior since without config switching the current position will be the form default position, which is what I will use).

eskwire commented:
Might be could test this with 4+ accounts for you.
